### ISSUE TITLE
FTL: The name 'AppDomain' does not exist in the current context (Source code: AppDomain)

### ISSUE THREAD

**Issue by riezebosch** [stryker-mutator/stryker-net#781]
**Describe the bug**

**Logs**
For us to best assist you in diagnosing the problem we would ask you to run stryker with the option `--log-file` and to attach the resulting logs to the issue.

This is broken? 
```
~ dotnet stryker -f stryker.log
Specify --help for a list of available options and commands.
```

```
Version: 0.13.0 (beta)

[14:34:44 INF] The project /Users/manuelriezebosch/git/Unmockable/Unmockable.Intercept/Unmockable.Intercept.csproj will be mutated
[14:34:46 INF] Started initial build using dotnet build
[14:34:48 INF] Initial build successful
[14:34:54 INF] Using testrunner VsTest
[14:34:54 INF] Total number of tests found: 48
[14:34:54 INF] Initial testrun started
[14:34:55 INF] Using 7536 ms as testrun timeout
[14:34:58 WRN] Stryker.NET encountered an compile error in Stryker.Core.InjectedHelpers.MutantControl.cs with message: The name 'AppDomain' does not exist in the current context (Source code: AppDomain)
[14:34:58 WRN] Safe Mode! Stryker will try to continue by rolling back all mutations in method. This should not happen, please report this as an issue on github with the previous error message.
[14:34:58 FTL] Stryker.NET could not compile the project after mutation. This is probably an error for Stryker.NET and not your project. Please report this issue on github with the previous error message.
[14:34:58 ERR] An error occurred during the mutation test run 
System.ApplicationException: Internal error due to compile error.
   at Stryker.Core.Compiling.RollbackProcess.RemoveMutantIfStatements(SyntaxTree originalTree, ICollection`1 diagnosticInfo, Boolean devMode)
   at Stryker.Core.Compiling.RollbackProcess.Start(CSharpCompilation compiler, ImmutableArray`1 diagnostics, Boolean devMode)
   at Stryker.Core.Compiling.CompilingProcess.TryCompilation(MemoryStream ms, CSharpCompilation compilation, EmitResult previousEmitResult, Boolean devMode, Int32 retryCount)
   at Stryker.Core.Compiling.CompilingProcess.Compile(IEnumerable`1 syntaxTrees, MemoryStream ms, Boolean devMode)
   at Stryker.Core.MutationTest.MutationTestProcess.Mutate()
   at Stryker.Core.StrykerRunner.RunMutationTest(StrykerOptions options, IEnumerable`1 initialLogMessages)
[14:34:58 INF] Time Elapsed 00:00:15.8082994
```
**Expected behavior**

**Desktop (please complete the following information):**
 - OS: Windows & MacOS
 - Type of project: core
 - Framework Version 3.0.100 (SDK)
 - Stryker Version: 0.13

**Additional context**
https://github.com/riezebosch/Unmockable

**Comment by Mobrockers** [comment]
Interesting.. Any idea why the AppDomain is not available to us in your project?

**Comment by richardwerkman** [comment]
`dotnet stryker -f stryker.log` does not work as you don't need to specify the log file name. Try:

`dotnet stryker -f`

**Comment by riezebosch** [comment]
Switching my test project to `netcoreapp2.1` makes it work but have to speficy `--language-version CSharp8` on the command line because of the _preview_ features I use.

**Comment by Mobrockers** [comment]
> The default coverage option `perTest` lets some mutants slip through.
> [fail.zip](https://github.com/stryker-mutator/stryker-net/files/3797762/fail.zip)
> [surviving.zip](https://github.com/stryker-mutator/stryker-net/files/3797765/surviving.zip)
> [100percent.zip](https://github.com/stryker-mutator/stryker-net/files/3797881/100percent.zip)

@dupdob Do you think these are the same issues you fixed in the work you're doing in combining non-overlapping mutations?

**Comment by riezebosch** [comment]
<img width="1350" alt="Screen Shot 2019-11-01 at 00 38 31" src="https://user-images.githubusercontent.com/1427334/68025863-3ed27600-fcae-11e9-9e96-ce1d126d37c6.png">
For the record 😎

**Comment by dupdob** [comment]
@Mobrockers: it has been a few months since I fixed this, so not sure when it was fixed and if it has been released or not; so it may be. Or it may be some other issue I have not identified yet.
@riezebosch: could you please provide the resulting status reports? or at least the mutants that slip through?

**Comment by riezebosch** [comment]
@dupdob checkout the attachments https://github.com/stryker-mutator/stryker-net/issues/781#issuecomment-548603635

**Comment by dupdob** [comment]
Thanks, I was hoping for the HTML reports, as the results are difficult to extract from the logs and I am not in a position to replicate the issue.

**Comment by riezebosch** [comment]
[default.zip](https://github.com/stryker-mutator/stryker-net/files/3800842/default.zip)
[off.zip](https://github.com/stryker-mutator/stryker-net/files/3800852/off.zip)

coverage analysis setting

**Comment by riezebosch** [comment]
Running stryker against `Unmockable.Wrap.Tests` keeps failing, also after specifying the language version or setting the target framework to `netcoreapp2.1`.

**Comment by dupdob** [comment]
Thanks for the reports. I just looked at them and saw nothing that could explain the significant difference between with or without coverage analysis. I have a couple of theories ⇒ next step is for me to reproduce the issue. I'll keep you posted.

**Comment by dupdob** [comment]
Quick update: I reproduce the coverage issue and I have started analysing it. The problem appears to happen during the coverage analysis phase, where some VsTest errors happen, invalidating data capture for the affected tests. Next step is to understand what triggers this errors and fix it.

**Comment by dupdob** [comment]
I have completed the analysis regarding the coverage situation.
## Issue: Significant discrepancy in mutation testing result when disabling coverage analysis on [Unmockable](https://github.com/riezebosch/Unmockable)
All mutants (37 at time of testing) are killed when turning coverage analysis off whereas 6 are uncovered and 3 survived with coverage analysis on.

## Cause: Unexpected impact of mutant control logic
**[Unmockable](https://github.com/riezebosch/Unmockable)** extensively uses the [System.Linq.Expressions.Expression](https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression?view=netframework-4.8) class in method signature. **StrykerMutator.Net** relies on an in-place mutation for expressions, using ternary operators to enable them when needed.
Therefore some mutated expressions are **not evaluated** at runtime, but simply passed as an expression tree to **Unmockable** code. 
As a result, mutant in **non evaluated expressions**:
- are invisible during coverage analysis  ⇒  flagged as not covered
- can't be selectively activated  ⇒  can't be tested 
- result mutated code that is active during every test sessions  ⇒ Stryker.Mutator results are wrong with or without coverage analysis.

To confirm that last statement, one simply has to manually introduce a mutant and run the code. For example, this mutant breaks no test and should be marked as survived 
```csharp
namespace Unmockable.Matchers
{
  internal static class ArgMatcherFactory
  {
    public static IArgumentMatcher? Create(Expression arg) => arg switch
      {
          MethodCallExpression call when call.Method.DeclaringType == typeof(Arg) => call.Method.Name switch
          {
              nameof(Arg.Ignore) => (IArgumentMatcher)new IgnoreArgument(),
              // here we have replaced Arguments.Single() with Arguments.SingleOrDefault()
              nameof(Arg.Where) => new LambdaArgument(call.Arguments.SingleOrDefault()),
              _ => throw new InvalidOperationException()
          },
          _ => null
      };
   }
}
```
But it is marked as *'not covered'* with coverage analysis and as *'killed'* when coverage analysis is off. Please note that having it marked as *'killed'* means it is very likely that other mutants are flagged as *'killed'* while they are actual survivors, invalidating the test session's results.



## Workaround
The only workaround is to use the ignore methods option (see #646) to disable mutation for each call to method expecting System.Linq.Expressions.Expression as parameter(s).

## Fix
At this time, there is no clear and fast approach to fix this issue. Leads are:
1. Identify offending method(s) and automatically disable related mutation. Can't be done as of now, as Stryker.Net does not implement the required call resolution logic to reliably identify those calls
2. Offer an option to only use if statements to control mutation logic. Difficult to do as it requires reliable identification of enclosing statements, it would result in bloated mutated assemblies due to the extra duplication and it is not universal (usage of expression body instead of statement)
3. A mixing **1** and **2** would limit the bloating (only use ifs when necessary) and handle the expression body problem by disabling mutation for those
4. Adjust mutant control logic to make it transparent for  System.Linq.Expressions.Expression. I have no idea how to achieve that, if it is possible.


## Other findings
This analysis clearly revealed that several logging messages where improperly formatted and/or where not providing the desired informations. I will open a PR that will fix all identified issues.

## Original issue
Now that I can reproduce the issue, I will have a look at the problem with 'AppDomain'

**Comment by Mobrockers** [comment]
As always awesome work!

**Comment by dupdob** [comment]
I have completed the analysis regarding the original issue with 'AppDomain'.

## Issue: Can't build some mutated assembly
Stryker.Mutator stops when building the mutated assembly with the following error message:
```
The name 'AppDomain' does not exist in the current context (Source code: AppDomain)
```

## Cause: No support for AppDomain from the project framework
Stryker.Net injects some custom code in the mutated assembly to implement mutant control features. Specifically this code uses AppDomain's events to report test coverage. But NetStandard and NetCore do not offer support for AppDomain prior V 2.0.  ⇒ Mutated assembly require either a classic Net Framework or NetStandard >= 2.0.

## Workaround
One can consider upgrading the framework version now that NetStandard 2.0 is mainstream.
Alternatively one can upgrade the framework version for mutation testing purposes.
*Note: This fix has been tested with success on Unmockable.*

## Fix
This issue can be fixed in StrykerMutator with limited impact (no support for coverage analysis **when using DotNetTest** on framework versions lacking support for AppDomain).
Impact remains to be confirmed once fix is implemented.

**Comment by riezebosch** [comment]
Perhaps Stryker could specify a different target framework on build with low impact?

**Comment by dupdob** [comment]
I want to avoid this solution, has it may have other consequences

**Comment by dupdob** [comment]
Re: issue about framework version:
I have resolved compilation issues. Next step is to restore coverage capture for NetStandard <2.0

**Comment by dupdob** [comment]
hummm. I fixed the issue with Unmockable, but it turns out to be a degenerated case as no mutant can be generated. So I tried with NFluent for NetStandard 1.3. The mutated assembly can be built and appears correct (using Jetbrain's Dotnet Peek) but failed to load with VsTest for unclear reason

**Comment by dupdob** [comment]
Well, it turns out that this problem helped identify a bunch of other issues, some of them being more or less hard to fix.
Therefore I will:

1. Provide a PR that fixes this issue ('AppDomain'). Basically, this fix will provide support for mutation testing for NetCore <2.0 and NetStandard <2.0 projects, but without any coverage analysis capability
2. Implements some easy fix (some logging and versioning)
3. Open issues for remaining problems and shortcomings.

**Comment by Mobrockers** [comment]
Main problem fixed by #789 

The fix should be released shortly. This does not however fix https://github.com/stryker-mutator/stryker-net/issues/781#issuecomment-549447726 

as a workaround for that problem the linq mutator could be disabled or the individual methods accepting a System.Linq.Expressions.Expression parameter could be ignored.

A new issue will be created to investigate the problem but as of right now a fix does not seem likely as stryker does not have the required information available to make the decision to not place the mutation in the offending way.
